keys: Clean up `MakeRowSentinelKey` and `EnsureSafeSplitKey`

a6802739 opened the pull request:
Fixes #16344. @bdarnell

a6802739 committed:
storage: Limit use of EnsureSafeSplitKey

EnsureSafeSplitKey transforms the requested split key in a way that is
only appropriate in some contexts. Since this was called in AdminSplit
itself, many call sites had to manually reverse this
transformation (by calling MakeRowSentinelKey) in order to split at
the desired location. This was easy to get wrong (see #16008).

This commit removes EnsureSafeSplitKey from AdminSplit, calling it
instead in MVCCFindSplitKey. Callers are now responsible for providing
valid split keys directly, instead of a key which will be a valid
split key after being transformed by EnsureSafeSplitKey.

Fixes #16344

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/16649)
<!-- Reviewable:end -->

bdarnell commented:
This looks good, but the tests are all failing. I think you also need to remove the call to MakeRowSentinelKey in `keys/constants.go:247`: `SystemConfigSplitKey = MakeRowSentinelKey(TableDataMin)`

---

Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/ccl/sqlccl/restore.go, line 536 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16649#-KnCxhXORCl5kXM6tqNw:-KnCxhXORCl5kXM6tqNx:bsa04td) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ab0f1d24d5066eed831e770159ec6a4657e6721/pkg/ccl/sqlccl/restore.go#L536)):*
> ```Go
> 		// Pick the index such that it's 0 if len(splitPoints) == 1.
> 		splitIdx := len(splitPoints) / 2
> 		// AdminSplit requires that the key be a valid table key, which means
> ```

Now you can remove this comment too. `splitKey` is just `splitPoints[splitIdx]` and the same value can be used for both arguments to `AdminSplit`. 

---

*[pkg/storage/replica_command.go, line 2642 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16649#-KnCy4f0hfQz7wFyFV6d:-KnCy4f0hfQz7wFyFV6e:b-mkubfk) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ab0f1d24d5066eed831e770159ec6a4657e6721/pkg/storage/replica_command.go#L2642)):*
> ```Go
> 		}
> 
> 		// EnsureSafeSplitKey could have changed the key, so we must
> ```

Remove this section too. 

---

*[pkg/storage/engine/mvcc.go, line 2313 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16649#-KnCykcKW2Dt8XutbLqi:-KnCykcKW2Dt8XutbLqj:b-muwttf) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ab0f1d24d5066eed831e770159ec6a4657e6721/pkg/storage/engine/mvcc.go#L2313)):*
> ```Go
> 	}
> 
> 	// The key has been removed family ID now.
> ```

"The family ID has been removed from this key, making it a valid split point."

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16649#-:-KnD-q89MXUWajIEtEbc:b-idjt3y)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/storage/engine/mvcc.go, line 2308 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16649#-KnEQFcPuq4YsBPdNrgO:-KnEQFcPuq4YsBPdNrgP:bkgmanl) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ab0f1d24d5066eed831e770159ec6a4657e6721/pkg/storage/engine/mvcc.go#L2308)):*
> ```Go
> 	}
> 
> 	splitKey, err := keys.EnsureSafeSplitKey(bestSplitKey.Key)
> ```

I think this new code is the same as `return keys.EnsureSafeSplitKey(bestSplitKey.Key)`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16649)*
<!-- Sent from Reviewable.io -->

a6802739 commented:
@bdarnell  I found a lot of error logs like

```
E170706 19:41:38.251592 49696 storage/queue.go:657  [split,s1,r4/1:/{System/tse-Max}] unable to split [n1,s1,r4/1:/{System/tse-
Max}] at key "/Table/0": storage/replica_command.go:2658: cannot split range at key /Table/0
```

if we remove `MakeRowSentinelKey` for `SystemConfigSplitKey`, will the split be invalid?

bdarnell commented:
I think [this line](https://github.com/cockroachdb/cockroach/blob/985b5d76c3d94dfd59e371da1185faea473c37a5/pkg/storage/engine/mvcc.go#L2224) in IsValidSplitKey needs to be changed to use `>` instead of `>=`. We want to allow splits at either end of the NoSplitSpans, just not in the middle.

a6802739 commented:
@bdarnell, I found it will report `had 4 ranges at startup, expected 10` for tests in `test_server_shim.go`, I don't know what does this mean. 

Need some help here. :)

bdarnell commented:
The remaining issues are test-only problems. Two tests were using MakeRowSentinelKey where it is no longer necessary, and an assertion in sql/main_test.go was incorrectly reporting errors in many tests because it was using keys.Addr when it should not. I've added a commit that should fix these.

bdarnell commented:
I had to rebase to pick up some build changes, and in the meantime another test was added that needed to be updated. This is ready to go now.

tamird commented:
Could you expand the commit message to (1) include the issue reference and (2) describe the actual change?

---

Reviewed 1 of 5 files at r1, 12 of 12 files at r2.
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/ccl/sqlccl/restore.go, line 536 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16649#-Kp1RaSptByPIW0ldTLO:-Kp1RaSptByPIW0ldTLP:bcb0s6g) ([raw file](https://github.com/cockroachdb/cockroach/blob/68584f19d631d2ccb7f962eb9f427752a81e3e14/pkg/ccl/sqlccl/restore.go#L536)):*
> ```Go
> 		// Pick the index such that it's 0 if len(splitPoints) == 1.
> 		splitIdx := len(splitPoints) / 2
> 		splitKey := append([]byte(nil), splitPoints[splitIdx]...)
> ```

why is this copy needed?

---

*[pkg/keys/constants.go, line 247 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16649#-Kp1R4YuEwB1b4pVwyr2:-Kp1R4YuEwB1b4pVwyr3:b-92yjv2) ([raw file](https://github.com/cockroachdb/cockroach/blob/68584f19d631d2ccb7f962eb9f427752a81e3e14/pkg/keys/constants.go#L247)):*
> ```Go
> 	// SystemConfigSplitKey is the key to split at immediately prior to the
> 	// system config span. NB: Split keys need to be valid column keys.
> 	SystemConfigSplitKey = []byte(TableDataMin)
> ```

why the different type?

---

*[pkg/storage/client_split_test.go, line 165 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16649#-Kp1RFmG6wquZ4KMtZFK:-Kp1RFmG6wquZ4KMtZFL:b9cw1eu) ([raw file](https://github.com/cockroachdb/cockroach/blob/68584f19d631d2ccb7f962eb9f427752a81e3e14/pkg/storage/client_split_test.go#L165)):*
> ```Go
> 	rowKey := roachpb.Key(encoding.EncodeVarintAscending(append([]byte(nil), tableKey...), 1))
> 	rowKey = encoding.EncodeStringAscending(encoding.EncodeVarintAscending(rowKey, 1), "a")
> 	col1Key, _ := keys.EnsureSafeSplitKey(keys.MakeFamilyKey(append([]byte(nil), rowKey...), 1))
> ```

why ignore these errors?

---

*[pkg/storage/client_split_test.go, line 306 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16649#-Kp1RCRKnPH18tNhGSwE:-Kp1RCRKnPH18tNhGSwF:bik8gr6) ([raw file](https://github.com/cockroachdb/cockroach/blob/68584f19d631d2ccb7f962eb9f427752a81e3e14/pkg/storage/client_split_test.go#L306)):*
> ```Go
> func TestEnsureSafeSplitKeyOutOfBounds(t *testing.T) {
> 	defer leaktest.AfterTest(t)()
> 	t.Skip()
> ```

did you mean to leave this in?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16649#-:-Kp1RlQNjcaB7MrnRphf:b-x6s3y7)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 10 of 14 files reviewed at latest revision, 7 unresolved discussions.

---

*[pkg/ccl/sqlccl/restore.go, line 536 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16649#-Kp1RaSptByPIW0ldTLO:-Kp1XeFiC1tK4jbrzoOS:b-9bvz85) ([raw file](https://github.com/cockroachdb/cockroach/blob/68584f19d631d2ccb7f962eb9f427752a81e3e14/pkg/ccl/sqlccl/restore.go#L536)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

why is this copy needed?
</blockquote></details>

I don't think it is any more. Deleted.

---

*[pkg/keys/constants.go, line 247 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16649#-Kp1R4YuEwB1b4pVwyr2:-Kp1XgZFyLb76k-BK6_0:b39cdop) ([raw file](https://github.com/cockroachdb/cockroach/blob/68584f19d631d2ccb7f962eb9f427752a81e3e14/pkg/keys/constants.go#L247)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

why the different type?
</blockquote></details>

MakeRowSentinelKey returns a `[]byte`, so this is the same as it was before this PR. It looks like it should be either a roachpb.Key or roachpb.RKey, but it's used inconsistently today. Added a TODO here and on MakeRowSentinelKey to change to a more precise type. 

---

*[pkg/storage/client_split_test.go, line 165 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16649#-Kp1RFmG6wquZ4KMtZFK:-Kp1Ydz9i-mOzX_Lk0hn:b-x21pan) ([raw file](https://github.com/cockroachdb/cockroach/blob/68584f19d631d2ccb7f962eb9f427752a81e3e14/pkg/storage/client_split_test.go#L165)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

why ignore these errors?
</blockquote></details>

No good reason; added checks.

---

*[pkg/storage/client_split_test.go, line 306 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16649#-Kp1RCRKnPH18tNhGSwE:-Kp1YRyA5R6O5b32Hjra:b-gi3mzm) ([raw file](https://github.com/cockroachdb/cockroach/blob/68584f19d631d2ccb7f962eb9f427752a81e3e14/pkg/storage/client_split_test.go#L306)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

did you mean to leave this in?
</blockquote></details>

No, I meant to remove this whole test but forgot to push my last commit.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16649)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16649)*
<!-- Sent from Reviewable.io -->

tamird commented:
LGTM modulo commit message changes.

---

Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16649#-:-Kp1oZwGp93Gp2TgNn1V:b-vdkhxy)*
<!-- Sent from Reviewable.io -->

bdarnell committed:
keys: Remove MakeRowSentinelKey

The concept of "row sentinel key" has been obsolete since the
introduction of column families. Now that splits do not require a
column family suffix, most calls to this method can be removed, and
the few remaining cases where it is meaningful are replaced with an
explicit use of column family 0.

bdarnell commented:
Rewrote commit message and added another commit which removes MakeRowSentinelKey entirely.

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> mod typo in first commit message:

> MVCCFIndSplitKey

Should be `MVCCFindSplitKey`

---

Reviewed 15 of 15 files at r4.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[pkg/keys/keys.go, line 571 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16649#-Kp2MKEqET_TJHHDxngZ:-Kp2MKEqET_TJHHDxng_:bvljuyt) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf94595c660c97786d3f34c1dbdf30b90306c9c9/pkg/keys/keys.go#L571)):*
> ```Go
> 
> // SentinelFamilyID indicates that MakeFamilyKey should make a sentinel key.
> const SentinelFamilyID = 0
> ```

since there are at least two comments that reference special behaviour associated with this value, perhaps keep this constant is worthwhile, if only to have a place to anchor comments.

---

*[pkg/sql/sqlbase/rowfetcher.go, line 399 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16649#-Kp2LdkDs5rOsWHq88d5:-Kp2LdkDs5rOsWHq88d6:b835e6m) ([raw file](https://github.com/cockroachdb/cockroach/blob/e12446ef2534db864ceccf7c88e54c71d6c95d3d/pkg/sql/sqlbase/rowfetcher.go#L399)):*
> ```Go
> 	prettyKey = prettyKeyPrefix
> 
> 	// If this is the row sentinel (in the legacy pre-family format),
> ```

isn't this also the case for column family zero, which is encoded without this value?

---

*[pkg/sql/sqlbase/table.go, line 1351 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16649#-Kp2LoVGsj0YrNWb3XR9:-Kp2LoVGsj0YrNWb3XRA:bmdr6ys) ([raw file](https://github.com/cockroachdb/cockroach/blob/e12446ef2534db864ceccf7c88e54c71d6c95d3d/pkg/sql/sqlbase/table.go#L1351)):*
> ```Go
> 	}
> 
> 	// Index keys are considered "sentinel" keys in that they do not have a
> ```

this comment is somewhat incongruent now

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16649#-:-Kp2LF33V4KOWZ5l9o9g:brlz0xe)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 10 of 25 files reviewed at latest revision, 10 unresolved discussions.

---

*[pkg/keys/keys.go, line 571 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16649#-Kp2MKEqET_TJHHDxngZ:-Kp2aiNOvbsUOIRX7S-U:b-cw9787) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf94595c660c97786d3f34c1dbdf30b90306c9c9/pkg/keys/keys.go#L571)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

since there are at least two comments that reference special behaviour associated with this value, perhaps keep this constant is worthwhile, if only to have a place to anchor comments.
</blockquote></details>

But the two places that referred to it were doing so for two separate reasons: column zero was a special sentinel in the pre-family encoding. With column families it's just a normal family with a small optimization in its key encoding. 

---

*[pkg/sql/sqlbase/rowfetcher.go, line 399 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16649#-Kp2LdkDs5rOsWHq88d5:-Kp2aLbPtVqDVKr05zT2:b-tmfbhq) ([raw file](https://github.com/cockroachdb/cockroach/blob/e12446ef2534db864ceccf7c88e54c71d6c95d3d/pkg/sql/sqlbase/rowfetcher.go#L399)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

isn't this also the case for column family zero, which is encoded without this value?
</blockquote></details>

No, family zero has values (including the entire table by default). The family-aware format doesn't go through `processValueSingle` (it uses `processValueTuple` instead). 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16649)*
<!-- Sent from Reviewable.io -->

tamird commented:
Thanks, LGTM.

On Jul 14, 2017 20:21, "Ben Darnell" <notifications@github.com> wrote:

Review status: 10 of 25 files reviewed at latest revision, 10 unresolved
discussions.
------------------------------

*pkg/keys/keys.go, line 571 at r3
<https://reviewable.io:443/reviews/cockroachdb/cockroach/16649#-Kp2MKEqET_TJHHDxngZ:-Kp2aiNOvbsUOIRX7S-U:b-cw9787>
(raw file
<https://github.com/cockroachdb/cockroach/blob/cf94595c660c97786d3f34c1dbdf30b90306c9c9/pkg/keys/keys.go#L571>):*
*Previously, tamird (Tamir Duberstein) wrote…*

since there are at least two comments that reference special behaviour
associated with this value, perhaps keep this constant is worthwhile, if
only to have a place to anchor comments.

But the two places that referred to it were doing so for two separate
reasons: column zero was a special sentinel in the pre-family encoding.
With column families it's just a normal family with a small optimization in
its key encoding.
------------------------------

*pkg/sql/sqlbase/rowfetcher.go, line 399 at r4
<https://reviewable.io:443/reviews/cockroachdb/cockroach/16649#-Kp2LdkDs5rOsWHq88d5:-Kp2aLbPtVqDVKr05zT2:b-tmfbhq>
(raw file
<https://github.com/cockroachdb/cockroach/blob/e12446ef2534db864ceccf7c88e54c71d6c95d3d/pkg/sql/sqlbase/rowfetcher.go#L399>):*
*Previously, tamird (Tamir Duberstein) wrote…*

isn't this also the case for column family zero, which is encoded without
this value?

No, family zero has values (including the entire table by default). The
family-aware format doesn't go through processValueSingle (it uses
processValueTuple instead).
------------------------------

*Comments from Reviewable
<https://reviewable.io:443/reviews/cockroachdb/cockroach/16649>*

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/cockroachdb/cockroach/pull/16649#issuecomment-315495702>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/ABdsPL0FUyihZFRwPw3DgdC_5SQ5nP0Uks5sOAYVgaJpZM4OAxzO>
.

a6802739 commented on pkg/cli/cli_test.go:
```diff
@@ -402,9 +402,9 @@ func Example_ranges() {
 	// /System/"tse"-"ranges3" [4]
 	// 	0: node-id=1 store-id=1
-	// "ranges3"-/Table/0 [11]
+	// "ranges3"-/Table/SystemConfigSpan/Start [11]
```
@bdarnell, Why we should change here?

bdarnell replied:
There's a special case in the key pretty-printer for the start of the system config span (which would otherwise look like `/Table/0/0`). We try to split at that key, but before this change, EnsureSafeSplitKey changed the split key to `/Table/0`. Now the split point correctly shows that it is `/Table/0/0` or its alias `/Table/SystemConfigSpan/Start`.

a6802739 commented on pkg/sql/main_test.go:
```diff
@@ -155,12 +155,8 @@ func checkEndTransactionTrigger(args storagebase.FilterArgs) *roachpb.Error {
-		}
-		if bytes.Compare(keyAddr, keys.SystemConfigSpan.Key) >= 0 &&
-			bytes.Compare(keyAddr, keys.SystemConfigSpan.EndKey) < 0 {
+		if bytes.Compare(span.Key, keys.SystemConfigSpan.Key) >= 0 &&
```
@bdarnell , I don't understand why we couldn't use `keys.Addr` here. Could you please explain here?

bdarnell replied:
We want to set the system config trigger for any *data* records in the system config span. We can determine that by looking at the keys alone. Using keys.Addr finds any records that live on the same range as the system config span, including range descriptors and transaction records. This bug was revealed when the split points changed so that range descriptors fall into this range.
